### PR TITLE
Remove `Default` bound on `AccountId` types under the xcm directory

### LINKED ISSUE #4687: Implications of Using `AccountId::default()` as parent `AccountId` for Location-Conversion
Given a note on the default behaviour of `AccountId32`

Quote:
> We recently discovered that anyone can sign transactions in Substrate with a zero account (AccountId::default()). The reason behind this is that cryptographically a private key of zero makes to an actual complete curve in sr25519 and ed255519. This is coming directly from schnorrkel.

Resolved for substrate via https://github.com/paritytech/substrate/pull/10403

Aren't the following lines a bit dangerous, given that many parachains will most likely plug in the normal `AccountId32` type there?

https://github.com/paritytech/polkadot/blob/5a7004f22b43e46e73ed36c87c9159283e025a56/xcm/xcm-builder/src/location_conversion.rs#L40-L59

### PR DESCRIPTION
**Pull request by KiChjang**
Fixes #4687.

cumulus companion: paritytech/cumulus#901

### CODE REVIEW

File: xcm/xcm-builder/src/location_conversion.rs [hunk 1]
```diff
@@ -35,22 +35,26 @@ impl<Network: Get<NetworkId>, AccountId: From<[u8; 32]> + Into<[u8; 32]> + Clone
...
+/// for `AccountId`.
+pub struct ParentIsAllZeroes<AccountId>(PhantomData<AccountId>);
+impl<AccountId: Decode + Eq + Clone> Convert<MultiLocation, AccountId>
+	for ParentIsAllZeroes<AccountId>
 {
 	fn convert_ref(location: impl Borrow<MultiLocation>) -> Result<AccountId, ()> {
 		if location.borrow().contains_parents_only(1) {
-			Ok(AccountId::default())
+			Ok(AccountId::decode(&mut TrailingZeroInput::zeroes())
+				.expect("infinite length input; no invalid inputs for type; qed"))
```
Comment (bkchr) [inline xcm/xcm-builder/src/location_conversion.rs:47]: For what is this account being used?

Comment (KiChjang) [reply xcm/xcm-builder/src/location_conversion.rs:47]: It looks like it is only used in testing, in cases where the test wants to convert a message originating from a parent location to just the zero account, i.e. a "don't care but signed" account.

As such, there's not really a huge issue leaving it as `ParentIsDefault`, other than to keep in line with the `Default` bound removal on `AccountId`.

Comment (shawntabrizi) [reply xcm/xcm-builder/src/location_conversion.rs:47]: I think we should remove this trait from the public api and make it test only.

Comment (xlc) [reply xcm/xcm-builder/src/location_conversion.rs:47]: I think default account is always zero account.

Comment (bkchr) [reply xcm/xcm-builder/src/location_conversion.rs:47]: Yeah `Default == Zero`.

I'm also for removing this or doing it like @xlc proposed.

**Comment by bkchr** [comment]
@KiChjang what is the status on this? I need this for my Substrate pr that removes the `Default` for `AccountId32`

**Comment by KiChjang** [comment]
@bkchr I'll need to figure out what the parent accounts are for Statemint/e, currently in the test networks I still have them set to the zero account, and it is definitely not going to be the zero account on production networks.

**Comment by shawntabrizi** [comment]
@KiChjang you could use something like `hash("parent_parachain")` or `hash("polkadot/kusama")` or something like that. Otherwise speak to Gav to finalize an answer.

File: xcm/xcm-builder/src/currency_adapter.rs [hunk 2]
```diff
@@ -52,22 +52,26 @@ impl From<Error> for XcmError {
...
 ///
 /// /// Our chain's account id.
 /// type AccountId = sp_runtime::AccountId32;
 ///
 /// /// Our relay chain's location.
 /// parameter_types! {
 ///     pub RelayChain: MultiLocation = Parent.into();
-///     pub CheckingAccount: AccountId = Default::default();
+///     pub CheckingAccount: AccountId =
+///         AccountId::decode(&mut TrailingZeroInput::zeroes()).unwrap();
```
Comment (bkchr) [inline xcm/xcm-builder/src/currency_adapter.rs:68]: Do we really want to have such "hacks" in the docs?

**Review by bkchr — APPROVED** [review]

File: xcm/xcm-builder/src/currency_adapter.rs (hunk 2 above)
Comment (KiChjang) [reply xcm/xcm-builder/src/currency_adapter.rs:68]: I don't know what else you can put into here, you want me to do the same thing with `PalletId` instead?

Comment (KiChjang) [reply xcm/xcm-builder/src/currency_adapter.rs:68]:
```suggestion
///     pub CheckingAccount: AccountId = PalletId(*b"checking").into_account();
```

**Review by shawntabrizi — APPROVED** [review]

**Review by bkchr — APPROVED** [review]

**Comment by KiChjang** [comment]
bot merge

**Pull request merged**